test/Infinispan.HotRod.Tests/CounterTest.cs

```diff
@@ -0,0 +1,362 @@
+            var cc = new CounterConfiguration(initialValue, 0, 200, 8, CounterType.BOUNDED_STRONG, Storage.VOLATILE);
+            rcm.DefineCounter(counterName, cc);
+            var counter = rcm.GetStrongCounter(counterName);
+            int val = counter.GetValue();
+            var name = counter.GetName();
+            Assert.AreEqual(initialValue, counter.GetValue());
+            rcm.Remove(counterName);
+        }
+
+        [Test]
+        public void ConfigurationStrongTest()
+        {
+            string counterName = "strongTestConfiguration";
+            var rcm = remoteManager.GetCounterManager();
+            var cc = new CounterConfiguration(0, -256, 65536, 64, CounterType.BOUNDED_STRONG, Storage.VOLATILE);
+            rcm.DefineCounter(counterName, cc);
+            var counter = rcm.GetStrongCounter(counterName);
+            ICounterConfiguration retCc = counter.GetConfiguration();
+            Assert.AreEqual(cc.GetLowerBound(), retCc.GetLowerBound());
+            Assert.AreEqual(cc.GetUpperBound(), retCc.GetUpperBound());
+            Assert.AreEqual(0, retCc.GetConcurrencyLevel());
+            Assert.AreEqual(cc.GetType(), retCc.GetType());
+            Assert.AreEqual(cc.GetStorage(), retCc.GetStorage());
+            Assert.AreEqual(counter.GetName(), counterName);
+            rcm.Remove(counterName);
+
+            string counterName1 = "strongTestConfiguration1";
+            var cc1 = new CounterConfiguration(0, -100, 100, 1, CounterType.UNBOUNDED_STRONG,
+                        Storage.PERSISTENT);
+            rcm.DefineCounter(counterName1, cc1);
+            counter = rcm.GetStrongCounter(counterName1);
+            ICounterConfiguration retCc1 = counter.GetConfiguration();
+            Assert.AreEqual(0, retCc1.GetLowerBound());
+            Assert.AreEqual(0, retCc1.GetUpperBound());
+            Assert.AreEqual(0, retCc1.GetConcurrencyLevel());
+            Assert.AreEqual(cc1.GetType(), retCc1.GetType());
+            Assert.AreEqual(cc1.GetStorage(), retCc1.GetStorage());
+            Assert.AreEqual(counter.GetName(), counterName1);
+            rcm.Remove(counterName1);
+        }
+
+        [Test]
+        public void RemoveStrongTest()
+        {
+            string counterName = "strongTestRemove";
+            int initialValue = 5;
+            var rcm = remoteManager.GetCounterManager();
+            var cc = new CounterConfiguration(initialValue, -100, 200, 8, CounterType.UNBOUNDED_STRONG, Storage.VOLATILE);
+            rcm.DefineCounter(counterName, cc);
+            var counter = rcm.GetStrongCounter(counterName);
+            Assert.AreEqual(counter.AddAndGet(100), 105);
+            rcm.Remove(counterName);
+            Assert.AreEqual(counter.GetValue(), 5);
+            Assert.AreEqual(counter.AddAndGet(-100), -95);
+            rcm.Remove(counterName);
+            Assert.AreEqual(counter.GetValue(), 5);
+            rcm.Remove(counterName);
```

Comment:
It would be also good to test that the Listener was removed upon rcm.Remove(counterName) - as long as it's true :) Maybe as a separate test. See my other question about the difference between rcm.Remove and counter.Remove().

Reply:
By design the listener should not be removed, if the counter is accessed again is recreated and the event flow from the server resumed